### PR TITLE
Geomodel

### PR DESCRIPTION
**Pull request by tzufgoogle**
"""Simple text classification example.
This script trains a text classifier on a dataset containing paragraphs that
are linked to s2cells. 

A sentence is encoded with an BERT and a binary prediction is made from CLS 
token embedding using an MLP.

Example command line call:
$ bazel-bin/cabby/model/text/s2cellid_prediction/model_trainer \
  --data_dir ~/data/wikigeo/pittsburgh  \
  --dataset_dir ~/model/wikigeo/dataset/pittsburgh \
  --region Pittsburgh \ 
  --s2_level 12 \
  --output_dir ~/tmp/output\

### CODE REVIEW

File: cabby/geo/util.py [hunk 1]
```diff
@@ -331,6 +382,77 @@ def check_if_geometry_in_polygon(geometry: Any, poly: Polygon) -> Polygon:
...
+  E.g, of string '(40.715865, -74.037258)'.
+  Arguments:
+    coord: A lat-lng coordinate to be converted to a tuple.
+  Returns:
+    A tuple of lat-lng.
+  '''
+  list_coords_str = coord_str.replace("(", "").replace(")", "").split(',')
+  coord = list(map(float, list_coords_str))
+
+  return (coord[0], coord[1])
```
Comment (sayalikulkarni) [inline cabby/geo/util.py:411]: I know that this is tricky - but some libraries use latitude-longitude versus some specify longitude,latitude. Is there a way to unify this? e.g L396 versus L411 are confusing unless we see that one uses shapely.Point
Consider using a named tuple

File: cabby/geo/util.py [hunk 2]
```diff
@@ -331,6 +382,77 @@ def check_if_geometry_in_polygon(geometry: Any, poly: Polygon) -> Polygon:
...
+
+
+def get_cell_center_from_s2cellids(
+    s2cell_ids: Sequence[int]) -> Sequence[Tuple]:
+  """Returns the center latitude and longitude of s2 cells.
+  Args:
```
Comment (sayalikulkarni) [inline cabby/geo/util.py:417]: 'Arguments' to be consistent?

File: cabby/geo/util.py [hunk 3]
```diff
@@ -331,6 +382,77 @@ def check_if_geometry_in_polygon(geometry: Any, poly: Polygon) -> Polygon:
...
+  for s2cellid in s2cell_ids:
+    s2_latlng = s2.S2CellId(int(s2cellid)).ToLatLng()
+    lat = s2_latlng.lat().degrees()
+    lng = s2_latlng.lng().degrees()
+    prediction_coords.append([lat, lng])
+  return np.array(prediction_coords)
+
+def project_point_in_segment(line_segment: LineString, point: Point):
+  """Projects point to line and check if the point projected is in a segment. 
+  Args:
```
Comment (sayalikulkarni) [inline cabby/geo/util.py:433]: same here

File: cabby/geo/util.py [hunk 4]
```diff
@@ -331,6 +382,77 @@ def check_if_geometry_in_polygon(geometry: Any, poly: Polygon) -> Polygon:
...
+    A tuple of lat-lng.
+  '''
+  list_coords_str = coord_str.replace("(", "").replace(")", "").split(',')
+  coord = list(map(float, list_coords_str))
+
+  return (coord[0], coord[1])
+
+
+def get_cell_center_from_s2cellids(
+    s2cell_ids: Sequence[int]) -> Sequence[Tuple]:
```
Comment (sayalikulkarni) [inline cabby/geo/util.py:415]: int64?

File: cabby/geo/util.py [hunk 5]
```diff
@@ -331,6 +382,77 @@ def check_if_geometry_in_polygon(geometry: Any, poly: Polygon) -> Polygon:
...
+    line_segment: The line segment.
+    point: The point to be projected on the line.
+  Returns:
+    1 if the projected point is in the segment and 0 if inot.
+  """
+
+  point = np.array(point.coords[0])
+
+  line_point_1 = np.array(line_segment.coords[0])
+  line_point_2 = np.array(line_segment.coords[len(line_segment.coords)-1])
```
Comment (sayalikulkarni) [inline cabby/geo/util.py:443]: why -1? what is the unit considered?

File: cabby/geo/util.py [hunk 6]
```diff
@@ -331,6 +382,77 @@ def check_if_geometry_in_polygon(geometry: Any, poly: Polygon) -> Polygon:
...
+
+  diff = line_point_2 - line_point_1
+  diff_norm = diff/np.linalg.norm(diff, 2)
+
+  projected_point = line_point_1 + diff_norm * \
+    np.dot(point - line_point_1, diff_norm)
+
+  projected_point = Point(projected_point)
+
+  return 1 if line_segment.distance(projected_point) < 1e-8 else 0
```
Comment (sayalikulkarni) [inline cabby/geo/util.py:453]: can rewrite this as:
return (line_segment.distance(projected_point) < 1e-8)

File: cabby/geo/util.py [hunk 7]
```diff
@@ -331,6 +382,77 @@ def check_if_geometry_in_polygon(geometry: Any, poly: Polygon) -> Polygon:
...
+    prediction_coords.append([lat, lng])
+  return np.array(prediction_coords)
+
+def project_point_in_segment(line_segment: LineString, point: Point):
+  """Projects point to line and check if the point projected is in a segment. 
+  Args:
+    line_segment: The line segment.
+    point: The point to be projected on the line.
+  Returns:
+    1 if the projected point is in the segment and 0 if inot.
```
Comment (sayalikulkarni) [inline cabby/geo/util.py:437]: change the return to a bool instead of 0-1

File: cabby/model/text/region_prediction/model_trainer.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+# Required flags.
+flags.mark_flag_as_required("data_dir")
+flags.mark_flag_as_required("output_dir")
+
+
+def main(argv):
+
+  device = torch.device(
+    'cuda') if torch.cuda.is_available() else torch.device('cpu')
```
Comment (sayalikulkarni) [inline cabby/model/text/region_prediction/model_trainer.py:81]: did you train on cpu?!

File: cabby/model/text/region_prediction/model_trainer.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+  train.train_model(
+    model=model,
+    device=device,
+    num_epochs=FLAGS.num_epochs,
+    optimizer=optimizer,
+    train_loader=train_loader,
+    valid_loader=valid_loader,
+    file_path=FLAGS.output_dir, 
+    eval_every=len(train_dataset) // 10
```
Comment (sayalikulkarni) [inline cabby/model/text/region_prediction/model_trainer.py:110]: prefer to eval every _static_ number of steps so even if the training set gets really large, we see the intermediate eval epochs.
eval_every=100 or 50?

File: cabby/model/text/region_prediction/train.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+        # Validation loop.
+        for batch in valid_loader:
+          input_ids = batch['input_ids'].to(device)
+          attention_mask = batch['attention_mask'].to(device)
+          labels = batch['labels'].to(device)
+          loss, logits = model(
+            input_ids, attention_mask=attention_mask, labels=labels)
+          valid_running_loss += loss.item()
+          topv, topi = logits.squeeze().topk(1)
+          topi = topi.squeeze().detach().cpu().numpy()
```
Comment (sayalikulkarni) [inline cabby/model/text/region_prediction/train.py:137]: does this work on a gpu?

File: cabby/model/text/region_prediction/train.py [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+
+      # Evaluation step.
+      if global_step % eval_every == 0:
+        model.eval()
+        total, correct = 0, 0
+
+        # Validation loop.
+        for batch in valid_loader:
+          input_ids = batch['input_ids'].to(device)
+          attention_mask = batch['attention_mask'].to(device)
```
Comment (sayalikulkarni) [inline cabby/model/text/region_prediction/train.py:131]: do you need a epsilon to avoid NaN when computing the logits with the attention mask?

File: cabby/model/text/s2cellid_prediction/BUILD [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    name = 'dataset',
+    srcs = ['dataset.py'],
+    deps = [
+          "//cabby/geo:util",
+          "//cabby/geo/map_processing:map_structure",
+
+    ],
+)
+
+
```
Comment (sayalikulkarni) [inline cabby/model/text/s2cellid_prediction/BUILD:16]: too many empty lines

**Review by sayalikulkarni — CHANGES_REQUESTED** [review]
One high level input - there seem to be too many models in the same PR. Suggest that you split them next time into one PR per model so it is easier to review :) 
It would also help to refactor out certain common code like the training loops, eval metrics etc. if any.

**Comment by tzufgoogle** [comment]
Build, run and tests working.

**Review by sayalikulkarni — APPROVED** [review]
this looks good to go...
